torch/cuda/amp/amp_scaler.py

```diff
@@ -0,0 +1,292 @@
+        """
+        if not self._enabled:
+            return outputs
+
+        def apply_scale(val):
+            if isinstance(val, torch.Tensor):
+                assert val.is_cuda
+                return val * self._scale
+            elif isinstance(val, container_abcs.Iterable):
+                return type(val)(apply_scale(v) for v in val)
+            else:
+                raise ValueError("outputs must be a Tensor or an iterable of Tensors")
+
+        return apply_scale(outputs)
+
+    def _unscale_grads(self, optimizer, rscale, found_inf, allow_fp16=False):
+        for group in optimizer.param_groups:
+            for param in group["params"]:
+                if param.grad is not None:
+                    if (not allow_fp16) and param.grad.dtype == torch.float16:
+                        raise ValueError("Attempting to unscale FP16 gradients.  If you want to check for "
+                                         "infs/nans without unscaling, use optimizer.check_infs() instead.")
+                    else:
+                        torch._amp_unscale_inf_check_(param.grad, rscale, found_inf)
+
+    def unscale(self, optimizer):
+        """
+        Divides ('unscales') the optimizer's gradient tensors by the scale factor.
+
+        If :meth:`unscale` is not called explicitly then gradients will be
+        automatically unscaled during :meth:`step`.
+
+        :meth:`unscale` can be called explicitly to manipulate a network's
+        gradients after backward and before :meth:`step`.
+
+        Simple example, using :meth:`unscale` to enable clipping of unscaled gradients::
+
+            ...
+            scaler.scale(loss).backward()
+            scaler.unscale(optimizer)
+            torch.nn.utils.clip_grad_norm_(model.parameters(), max_norm)
+            scaler.step(optimizer)
+            scaler.update()
+
+        If this instance of :class:`AmpScaler` is not enabled, :meth:`unscale` is a no-op.
+
+        Arguments:
+            optimizer (torch.optim.Optimizer):  Optimizer that owns the gradients to be unscaled.
+
+        .. note::
+            :meth:`unscale` does not incur a CPU-GPU sync.
+
+        .. warning::
+            :meth:`unscale` should only be called once per optimizer per step,
+            and only after _all_ gradients for that optimizer's owned parameters
+            have been accumulated.
+        """
+        if not self._enabled:
+            return
+
+        optimizer_state = self._per_optimizer_states[id(optimizer)]
+
+        if "unscaled" in optimizer_state:
+            raise RuntimeError("unscale() has already been called on this optimizer this iteration.")
+
+        # FP32 division can be imprecise for certain compile options, so we carry out the reciprocal in FP64.
+        rscale = self._scale.double().reciprocal().float()
+        found_inf = torch.full((1,), 0.0, dtype=torch.float32, device="cuda")
```

Comment:
Do you need to implement a .to() method if people are working on multiple gpus? What would your advice be? One AmpScaler per device?

Reply:
If the user requests that I scale loss or output that happens to be on a different device from anything we've seen so far, i'll lazily create a copied scale tensor on that device.